app/src/main/java/de/grobox/transportr/trips/search/DirectionsFragment.kt

```diff
@@ -0,0 +1,274 @@
+            fromLocation.setWorkLocation(it)
+            viaLocation.setWorkLocation(it)
+            toLocation.setWorkLocation(it)
+        })
+        viewModel.locations.observe(viewLifecycleOwner, Observer {
+            if (it == null) return@Observer
+            fromLocation.setFavoriteLocations(it)
+            viaLocation.setFavoriteLocations(it)
+            toLocation.setFavoriteLocations(it)
+        })
+        viewModel.fromLocation.observe(viewLifecycleOwner, Observer {
+            fromLocation.setLocation(it)
+            if (it != null) toLocation.requestFocus()
+        })
+        viewModel.viaLocation.observe(viewLifecycleOwner, Observer { viaLocation.setLocation(it) })
+        viewModel.toLocation.observe(viewLifecycleOwner, Observer { toLocation.setLocation(it) })
+        viewModel.isDeparture.observe(viewLifecycleOwner, Observer<Boolean> { this.onIsDepartureChanged(it) })
+        viewModel.isExpanded.observe(viewLifecycleOwner, Observer<Boolean> { this.onViaVisibleChanged(it) })
+        viewModel.calendar.observe(viewLifecycleOwner, Observer { this.onCalendarUpdated(it) })
+        viewModel.findGpsLocation.observe(viewLifecycleOwner, Observer { this.onFindGpsLocation(it) })
+        viewModel.isFavTrip.observe(viewLifecycleOwner, Observer { this.onFavStatusChanged(it) })
+        viewModel.products.observe(viewLifecycleOwner, Observer { this.onProductsChanged(it) })
+
+        favIcon.setOnClickListener { viewModel.toggleFavTrip() }
+
+        timeBackground.setOnClickListener {
+            if (viewModel.calendar.value == null) throw IllegalStateException()
+            val fragment = TimeDateFragment.newInstance(viewModel.calendar.value!!, viewModel.isDeparture.value!!)
+            fragment.setTimeDateListener(viewModel)
+            fragment.show(activity!!.supportFragmentManager, TimeDateFragment.TAG)
+        }
+        timeBackground.setOnLongClickListener {
+            viewModel.resetCalender()
+            true
+        }
+
+        productsIcon.setOnClickListener {
+            activity?.let { a ->
+                ProductDialogFragment().show(a.supportFragmentManager, ProductDialogFragment.TAG)
+            }
+        }
+        swapIcon.setOnClickListener { swapLocations() }
+        viaIcon.setOnClickListener { viewModel.toggleIsExpanded() }
+
+        TooltipCompat.setTooltipText(productsIcon, getString(R.string.action_choose_products))
+        TooltipCompat.setTooltipText(swapIcon, getString(R.string.action_switch_locations))
+        TooltipCompat.setTooltipText(viaIcon, getString(R.string.action_navigation_expand))
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putSerializable(DATE, viewModel.calendar.value)
+        outState.putBoolean(EXPANDED, viewModel.isExpanded.value ?: false)
+        outState.putBoolean(DEPARTURE, viewModel.isDeparture.value ?: true)
```

Comment:
As a future optimization we can look into https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate

Reply:
Sounds interesting indeed.

Reply:
Moved to #667.